### PR TITLE
rustdoc: Cleanup various `clean` types

### PR DESCRIPTION
**Pull request by camelid**
Cleanup various `clean` types.

### CODE REVIEW

**Review by jyn514 — COMMENTED** [review]
Got about halfway through and ran out of time - this is amazing work ❤️ not sure why it's marked as a draft, but feel free to ignore the comments until you're happy with it :)

File: src/librustdoc/clean/auto_trait.rs [hunk 1]
```diff
@@ -353,7 +353,7 @@ impl<'a, 'tcx> AutoTraitFinder<'a, 'tcx> {
                     let (poly_trait, output) =
                         (data.0.as_ref().expect("as_ref failed").clone(), data.1.as_ref().cloned());
                     let new_ty = match poly_trait.trait_ {
-                        Type::ResolvedPath { ref path, ref did, ref is_generic } => {
+                        Type::ResolvedPath { ref path, ref did } => {
```
[the remark this answers is not in the document]
Comment (camelid) [reply src/librustdoc/clean/auto_trait.rs]: Yeah, but I was trying to minimize diff noise :)

File: src/librustdoc/clean/mod.rs [hunk 2]
```diff
@@ -1753,9 +1752,8 @@ impl Clean<Variant> for hir::VariantData<'_> {
 impl Clean<Path> for hir::Path<'_> {
     fn clean(&self, cx: &mut DocContext<'_>) -> Path {
         Path {
-            global: self.is_global(),
             res: self.res,
-            segments: if self.is_global() { &self.segments[1..] } else { &self.segments }.clean(cx),
+            segments: self.segments.clean(cx),
```
[the remark this answers is not in the document]
Comment (camelid) [reply src/librustdoc/clean/mod.rs]: I thought `kw::PathRoot`, but it may not be because otherwise I'd expect tests to be failing. I'll have to look into it.

File: src/librustdoc/clean/mod.rs [hunk 3]
```diff
@@ -1751,10 +1748,7 @@ impl Clean<Variant> for hir::VariantData<'_> {
 
 impl Clean<Path> for hir::Path<'_> {
     fn clean(&self, cx: &mut DocContext<'_>) -> Path {
-        Path {
-            res: self.res,
-            segments: self.segments.clean(cx),
-        }
+        Path { res: self.res, segments: self.segments.clean(cx) }
     }
```
[the remark this answers is not in the document]
Comment (camelid) [reply src/librustdoc/clean/mod.rs:1733]: Fixed!

File: src/librustdoc/clean/types.rs [hunk 4]
```diff
@@ -2002,8 +2001,7 @@ impl Path {
     }
 
     crate fn whole_name(&self) -> String {
-        String::from(if self.global { "::" } else { "" })
-            + &self.segments.iter().map(|s| s.name.to_string()).collect::<Vec<_>>().join("::")
+        self.segments.iter().map(|s| s.name.to_string()).collect::<Vec<_>>().join("::")
```
[the remark this answers is not in the document]
Comment (camelid) [reply src/librustdoc/clean/types.rs]: Fixed, thanks!

[the remark this answers is not in the document]
Comment (camelid) [reply src/librustdoc/clean/types.rs]: Actually, I think `intersperse` might be *less* efficient in this case because it will be cloning the `::`, whereas `join` seems to not clone the separator. Perhaps we should just leave it as-is?

File: src/librustdoc/html/format.rs [hunk 5]
```diff
@@ -732,9 +732,9 @@ fn fmt_type<'cx>(
 
     match *t {
         clean::Generic(name) => write!(f, "{}", name),
-        clean::ResolvedPath { did, ref path, is_generic } => {
+        clean::ResolvedPath { did, ref path } => {
             // Paths like `T::Output` and `Self::Output` should be rendered with all segments.
-            resolved_path(f, did, path, is_generic, use_absolute, cx)
+            resolved_path(f, did, path, path.is_generic(), use_absolute, cx)
```
[the remark this answers is not in the document]
Comment (camelid) [reply src/librustdoc/html/format.rs]: You're looking at an old diff; I changed `path.is_generic()` to `false` in a later commit (the one we're nervous about ;).

[the remark this answers is not in the document]
Comment (camelid) [reply src/librustdoc/html/format.rs]: I should probably remove the comment above this call too.

File: src/librustdoc/clean/types.rs [hunk 6]
```diff
@@ -2020,7 +2020,7 @@ crate enum GenericArg {
 #[derive(Clone, PartialEq, Eq, Debug, Hash)]
 crate enum GenericArgs {
     AngleBracketed { args: Vec<GenericArg>, bindings: Vec<TypeBinding> },
-    Parenthesized { inputs: Vec<Type>, output: Option<Type> },
+    Parenthesized { inputs: Vec<Type>, output: Option<Box<Type>> },
```
[the remark this answers is not in the document]
Comment (jyn514) [reply src/librustdoc/clean/types.rs:2008]: I would prefer not to make this change without measuring it, and I'd like it to be a separate PR either way. If you do make a separate PR, could you add a rustc_assert_size call so we can see how much it changed?

**Comment by jyn514** [comment]
> It's mainly marked as draft because some of the changes need more investigation (e.g., #88379 (comment)).

Ok. Can you split those into a separate PR so we can land the rest? There are a lot of great changes here :) but it's an awfully large diff and it makes it harder to review the more subtle changes.

**Comment by camelid** [comment]
> > It's mainly marked as draft because some of the changes need more investigation (e.g., #88379 (comment)).
> 
> Ok. Can you split those into a separate PR so we can land the rest? There are a lot of great changes here :) but it's an awfully large diff and it makes it harder to review the more subtle changes.

Sure! That's another reason why I marked it as draft, since I expected it would need to be split up.

**Comment by camelid** [comment]
@bors rollup=never

Since this may improve performance.

File: src/librustdoc/clean/types.rs (hunk 4 above)
[the remark this answers is not in the document]
Comment (jyn514) [reply src/librustdoc/clean/types.rs]: :shrug: I think intersperse is easier to read, but I don't feel strongly, whatever you have currently is fine.

File: src/librustdoc/html/format.rs [hunk 7]
```diff
@@ -805,28 +805,17 @@ fn fmt_type<'cx>(
...
-                            cx,
-                        )
-                    }
-                }
-                _ => {
-                    primitive_link(f, clean::PrimitiveType::RawPointer, &format!("*{} ", m), cx)?;
-                    fmt::Display::fmt(&t.print(cx), f)
-                }
+
+            if matches!(**t, clean::Generic(_)) || t.is_generic() {
```
[the remark this answers is not in the document]
Comment (jyn514) [reply src/librustdoc/html/format.rs]: > is_generic() is weird. According to the docs for the field that is_generic() replaced, it records whether the path is for an associated type.

Oh, that's horrifying. Can you change the name to `is_associated_type`?

File: src/librustdoc/html/render/print_item.rs [hunk 8]
```diff
@@ -699,7 +699,7 @@ fn item_trait(w: &mut Buffer, cx: &Context<'_>, it: &clean::Item, t: &clean::Tra
                 clean::ResolvedPath { ref path, did, .. }
                 | clean::BorrowedRef {
                     type_: box clean::ResolvedPath { ref path, did, .. }, ..
-                } if !path.is_generic() => {
+                } => {
```
[the remark this answers is not in the document]
Comment (camelid) [reply src/librustdoc/html/render/print_item.rs]: Ok, I removed this change for now.

File: src/librustdoc/clean/types.rs (hunk 6 above)
[the remark this answers is not in the document]
Comment (camelid) [reply src/librustdoc/clean/types.rs]: This was made into a separate PR: #88522. Resolving thread...

File: src/librustdoc/html/format.rs (hunk 7 above)
[the remark this answers is not in the document]
Comment (camelid) [reply src/librustdoc/html/format.rs]: Ok, I changed it to `is_assoc_type` (to be shorter and more consistent with rustc/rustdoc's abbreviated naming scheme).

File: src/librustdoc/html/format.rs (hunk 5 above)
[the remark this answers is not in the document]
Comment (camelid) [reply src/librustdoc/html/format.rs]: I have since dropped the commit that removed `is_generic()`, so this place now has `resolved_path(..., path.is_assoc_ty())`. I think `resolved_path` and related functions need to be cleaned up a lot in general, but I think let's do that in a different PR so this one doesn't keep growing ;)

[the remark this answers is not in the document]
Comment (camelid) [reply src/librustdoc/html/format.rs]: Actually, it looks like I should be able to pretty easily clean up `resolved_path` such that the argument that `path.is_generic()` is used for is removed!

[the remark this answers is not in the document]
Comment (camelid) [reply src/librustdoc/html/format.rs]: ... or not; it looks like there's a single place where `use_absolute` is not `false` and not `None` :/

File: src/librustdoc/clean/mod.rs (hunk 2 above)
[the remark this answers is not in the document]
Comment (camelid) [reply src/librustdoc/clean/mod.rs]: It will be `kw::PathRoot`, which renders as `{{root}}`, but I haven't found a way to make rustdoc (erroneously) print out `{{root}}::foo::bar`.

**Comment by camelid** [comment]
@jyn514 Ok, I think this should be ready for another review. There are a few unresolved threads:

1. https://github.com/rust-lang/rust/pull/88379#discussion_r697455410
2. https://github.com/rust-lang/rust/pull/88379#discussion_r697463052
3. https://github.com/rust-lang/rust/pull/88379#discussion_r697465028

I think the second and third ones should probably be left for future PRs. So, the first one is the main thread to be discussed. Or, if there's anything else you notice or have questions about, let me know!

I can also try to split up this PR further if it would make it easier to review.

File: src/librustdoc/clean/mod.rs [hunk 9]
```diff
@@ -1298,31 +1292,32 @@ fn clean_qpath(hir_ty: &hir::Ty<'_>, cx: &mut DocContext<'_>) -> Type {
...
+                trait_,
             }
         }
         hir::QPath::TypeRelative(ref qself, ref segment) => {
             let ty = hir_ty_to_ty(cx.tcx, hir_ty);
             let res = if let ty::Projection(proj) = ty.kind() {
                 Res::Def(DefKind::Trait, proj.trait_ref(cx.tcx).def_id)
             } else {
-                Res::Err
+                bug!("clean: expected associated item")
```
Comment (camelid) [inline src/librustdoc/clean/mod.rs:1312]: I'll note that this panics on the following code:

```rust
#![feature(inherent_associated_types)]

pub struct Struct;

impl Struct {
    pub type AssocTy = usize;
    pub const AssocConst: Self::AssocTy = 42;
}
```

```
warning: the feature `inherent_associated_types` is incomplete and may not be safe to use and/or cause compiler crashes
 --> inherent-associated-types.rs:1:12
  |
1 | #![feature(inherent_associated_types)]
  |            ^^^^^^^^^^^^^^^^^^^^^^^^^
  |
  = note: `#[warn(incomplete_features)]` on by default
  = note: see issue #8995 <https://github.com/rust-lang/rust/issues/8995> for more information

error[E0223]: ambiguous associated type
 --> inherent-associated-types.rs:7:27
  |
7 |     pub const AssocConst: Self::AssocTy = 42;
  |                           ^^^^^^^^^^^^^ help: use fully-qualified syntax: `<Struct as Trait>::AssocTy`

error: internal compiler error: src/librustdoc/clean/mod.rs:1312:17: clean: expected associated item

thread 'rustc' panicked at 'Box<dyn Any>', compiler/rustc_errors/src/lib.rs:1147:9
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace

note: the compiler unexpectedly panicked. this is a bug.

note: we would appreciate a bug report: https://github.com/rust-lang/rust/issues/new?labels=C-bug%2C+I-ICE%2C+T-compiler&template=ice.md

note: rustc 1.56.0-nightly (517c28e42 2021-08-26) running on x86_64-apple-darwin

query stack during panic:
end of query stack
error: aborting due to 2 previous errors; 1 warning emitted

For more information about this error, try `rustc --explain E0223`.
```

However:

* A rustc error is emitted for this code, before rustdoc ICEs (I'm not very familiar with `inherent_associated_types`, so I'm not sure how the error would be fixed)
* This is an unstable and incomplete feature
* rustdoc panics on nightly too, though at a different spot (because this change caused rustdoc to fail early):

```
warning: the feature `inherent_associated_types` is incomplete and may not be safe to use and/or cause compiler crashes
 --> inherent-associated-types.rs:1:12
  |
1 | #![feature(inherent_associated_types)]
  |            ^^^^^^^^^^^^^^^^^^^^^^^^^
  |
  = note: `#[warn(incomplete_features)]` on by default
  = note: see issue #8995 <https://github.com/rust-lang/rust/issues/8995> for more information

error[E0223]: ambiguous associated type
 --> inherent-associated-types.rs:7:27
  |
7 |     pub const AssocConst: Self::AssocTy = 42;
  |                           ^^^^^^^^^^^^^ help: use fully-qualified syntax: `<Struct as Trait>::AssocTy`

thread 'rustc' panicked at 'attempted .def_id() on invalid res: Err', /rustc/5eacec9ec7e112a0de1011519a57c45586d58414/compiler/rustc_hir/src/def.rs:538:32
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace

error: internal compiler error: unexpected panic

note: the compiler unexpectedly panicked. this is a bug.

note: we would appreciate a bug report: https://github.com/rust-lang/rust/issues/new?labels=C-bug%2C+I-ICE%2C+T-compiler&template=ice.md

note: rustc 1.56.0-nightly (5eacec9ec 2021-08-28) running on x86_64-apple-darwin

query stack during panic:
end of query stack
error: aborting due to previous error; 1 warning emitted

For more information about this error, try `rustc --explain E0223`.
```

Comment (camelid) [reply src/librustdoc/clean/mod.rs:1312]: Ok, I opened a new PR for that: #88573. I'll drop the commit in this PR that changes the `Res::Err` to a `bug!` since I incorporated it into the other PR.

**Comment by camelid** [comment]
Rebased to fix conflicts.

**Comment by camelid** [comment]
It's probably best to review this commit-by-commit, with whitespace changes hidden.

**Comment by camelid** [comment]
Ok, I rebased to fix the merge conflicts, and as part of that, dropped the commits that were merged as part of #88810. 🎉 

Now I'm going to create a new "remix" part 2 PR :)

**Comment by camelid** [comment]
Rebased.

File: src/librustdoc/clean/inline.rs [hunk 10]
```diff
@@ -455,11 +455,20 @@ crate fn build_impl(
...
+    let mut stack: Vec<&Type> = vec![&for_];
+
+    if let Some(did) = trait_.as_ref().map(|t| t.res.def_id()) {
+        if tcx.get_attrs(did).lists(sym::doc).has_word(sym::hidden) {
+            return;
+        }
+    }
+    if let Some(generics) = trait_.as_ref().and_then(|t| t.generics()) {
+        stack.extend(generics);
+    }
```
Comment (jyn514) [inline src/librustdoc/clean/inline.rs]: Why this change?

File: src/librustdoc/clean/mod.rs [hunk 11]
```diff
@@ -1286,11 +1287,12 @@ fn clean_qpath(hir_ty: &hir::Ty<'_>, cx: &mut DocContext<'_>) -> Type {
                 res: Res::Def(DefKind::Trait, trait_def),
                 segments: trait_segments.clean(cx),
             };
+            register_res(cx, trait_path.res);
```
Comment (jyn514) [inline src/librustdoc/clean/mod.rs]: Does `hir::QPath::Resolved(None, _)` above also need `register_res` added?

File: src/librustdoc/clean/simplify.rs [hunk 12]
```diff
@@ -97,17 +97,13 @@ crate fn merge_bounds(
...
         };
-        let (did, path) = match trait_ref.trait_ {
-            clean::ResolvedPath { did, ref mut path, .. } => (did, path),
-            _ => return false,
-        };
         // If this QPath's trait `trait_did` is the same as, or a supertrait
         // of, the bound's trait `did` then we can keep going, otherwise
         // this is just a plain old equality bound.
-        if !trait_is_same_or_supertrait(cx, did, trait_did) {
+        if !trait_is_same_or_supertrait(cx, trait_ref.trait_.res.def_id(), trait_did) {
```
Comment (jyn514) [inline src/librustdoc/clean/simplify.rs]: maybe add a helper `def_id()` function to `trait_ref`?

File: src/librustdoc/clean/types.rs [hunk 13]
```diff
@@ -1409,7 +1409,8 @@ crate enum Type {
         name: Symbol,
         self_type: Box<Type>,
         self_def_id: Option<DefId>,
-        trait_: Box<Type>,
+        // FIXME: remove this `Box`; it's unnecessary
+        trait_: Box<Path>,
```
Comment (jyn514) [inline src/librustdoc/clean/types.rs]: Is this changed later in the PR? It seems like it wouldn't be too much effort.

File: src/librustdoc/clean/utils.rs [hunk 14]
```diff
@@ -156,39 +156,7 @@ pub(super) fn external_path(
     }
 }
 
-crate fn strip_type(ty: Type) -> Type {
-    match ty {
-        Type::ResolvedPath { path, did } => Type::ResolvedPath { path: strip_path(&path), did },
```
Comment (jyn514) [inline src/librustdoc/clean/utils.rs:162]: Wait, was this never called on anything besides `ResolvedPath` before ??

File: src/librustdoc/clean/utils.rs [hunk 15]
```diff
@@ -157,39 +156,7 @@ pub(super) fn external_path(
...
-            self_def_id,
-            self_type: Box::new(strip_type(*self_type)),
-            trait_: Box::new(strip_type(*trait_)),
-        },
-        _ => ty,
-    }
-}
-
-crate fn strip_path(path: &Path) -> Path {
+crate fn strip_path(path: Path) -> Path {
```
Comment (jyn514) [inline src/librustdoc/clean/utils.rs:159]:
```suggestion
/// Remove the generic arguments from a path.
crate fn strip_path(path: Path) -> Path {
```

File: src/librustdoc/html/render/mod.rs [hunk 16]
```diff
@@ -2378,14 +2387,7 @@ fn collect_paths_for_type(first_ty: clean::Type, cache: &Cache) -> Vec<String> {
...
         match ty {
-            clean::Type::ResolvedPath { did, .. } => {
-                let get_extern = || cache.external_paths.get(&did).map(|s| s.0.clone());
-                let fqp = cache.exact_paths.get(&did).cloned().or_else(get_extern);
-
-                if let Some(path) = fqp {
-                    out.push(path.join("::"));
-                }
-            }
+            clean::Type::ResolvedPath { did, .. } => process_path(did),
```
Comment (jyn514) [inline src/librustdoc/html/render/mod.rs:2374]: Why change this to a closure?

File: src/librustdoc/html/render/mod.rs [hunk 17]
```diff
@@ -2403,7 +2405,7 @@ fn collect_paths_for_type(first_ty: clean::Type, cache: &Cache) -> Vec<String> {
             }
             clean::Type::QPath { self_type, trait_, .. } => {
                 work.push_back(*self_type);
-                work.push_back(*trait_);
+                process_path(trait_.res.def_id());
```
Comment (jyn514) [inline src/librustdoc/html/render/mod.rs]: This changes the order the paths are processed - previously `self_type` was processed first, now `trait_` is. Does that change the behavior?

File: src/librustdoc/clean/inline.rs [hunk 18]
```diff
@@ -447,7 +447,7 @@ crate fn build_impl(
     };
     let polarity = tcx.impl_polarity(did);
     let trait_ = associated_trait.clean(cx);
-    if trait_.def_id() == tcx.lang_items().deref_trait() {
+    if trait_.as_ref().map(|t| t.def_id()) == tcx.lang_items().deref_trait() {
```
Comment (jyn514) [inline src/librustdoc/clean/inline.rs:450]: This doesn't seem like a very helpful change? I would rather keep GetDefId than add more churn like this.

File: src/librustdoc/clean/types.rs [hunk 19]
```diff
@@ -1368,60 +1366,65 @@ crate struct TraitAlias {
...
 crate enum Type {
-    /// Structs/enums/traits (most that would be an `hir::TyKind::Path`).
-    ResolvedPath {
-        path: Path,
-        did: DefId,
-    },
-    /// `dyn for<'a> Trait<'a> + Send + 'static`
+    /// A named type, which could be a trait.
+    ///
+    /// This is mostly Rustdoc's version of [`hir::Path`].
```
Comment (jyn514) [inline src/librustdoc/clean/types.rs:1378]:
```suggestion
    /// This is mostly Rustdoc's version of [`hir::Path`]. It has to be different because Rustdoc's `PathSegment` can contain cleaned generics.
```

File: src/librustdoc/clean/types.rs [hunk 20]
```diff
@@ -1372,56 +1372,61 @@ crate struct PolyTrait {
...
     DynTrait(Vec<PolyTrait>, Option<Lifetime>),
-    /// For parameterized types, so the consumer of the JSON don't go
-    /// looking for types which don't exist anywhere.
+    /// A type parameter.
     Generic(Symbol),
-    /// Primitives are the fixed-size numeric types (plus int/usize/float), char,
-    /// arrays, slices, and tuples.
+    /// A primitive (aka, builtin) type.
+    ///
+    /// This represents things like `i32`, `str`, etc.
```
Comment (jyn514) [inline src/librustdoc/clean/types.rs:1386]: Personally I wouldn't bother adding docs on this, since there's already docs on PrimitiveType.

File: src/librustdoc/clean/types.rs [hunk 21]
```diff
@@ -1372,56 +1372,61 @@ crate struct PolyTrait {
...
     Tuple(Vec<Type>),
+    /// A slice type (does *not* include the `&`): `[i32]`
     Slice(Box<Type>),
-    /// The `String` field is about the size or the constant representing the array's length.
+    /// An array type.
+    ///
+    /// The `String` field is a stringified version of the array's length parameter.
     Array(Box<Type>, String),
+    /// The never type: `!`
     Never,
```
Comment (jyn514) [inline src/librustdoc/clean/types.rs:1399]: This is already in `PrimitiveType`. Why is it duplicated here?

(ok if you're not sure, but if so please open an issue and leave a fixme here)

File: src/librustdoc/clean/types.rs [hunk 22]
```diff
@@ -1372,56 +1372,61 @@ crate struct PolyTrait {
...
+        /// FIXME: This is a hack that should be removed; see [this discussion][1].
+        ///
+        /// [1]: https://github.com/rust-lang/rust/pull/85479#discussion_r635729093
         self_def_id: Option<DefId>,
         trait_: Path,
     },
 
-    // `_`
+    /// A type that is inferred: `_`
     Infer,
```
Comment (jyn514) [inline src/librustdoc/clean/types.rs:1417]: When can this happen? Shouldn't all types be inferred by the time rustdoc gets to them?

The only uses I see are
```rust
src/librustdoc/clean/mod.rs
1295:                // Rustdoc handles `ty::Error`s by turning them into `Type::Infer`s.
1296:                ty::Error(_) => return Type::Infer,
1362:            // Rustdoc handles `TyKind::Err`s by turning them into `Type::Infer`s.
1363:            TyKind::Infer | TyKind::Err => Infer,
1758:                        hir::GenericArg::Infer(_inf) => GenericArg::Infer,
```
so maybe this should say "a type error" instead and be removed? Is it possible to get `TyKind::Infer` at all this late in analysis?

File: src/librustdoc/clean/types.rs [hunk 23]
```diff
@@ -120,8 +120,6 @@ crate struct Crate {
     crate module: Item,
     crate externs: Vec<ExternalCrate>,
     crate primitives: ThinVec<(DefId, PrimitiveType)>,
-    // These are later on moved into `CACHEKEY`, leaving the map empty.
-    // Only here so that they can be filtered through the rustdoc passes.
```
Comment (jyn514) [inline src/librustdoc/clean/types.rs:124]: Please leave the second line of the comment here; there's some quite tricky things going on around `external_traits`.

**Review by jyn514 — CHANGES_REQUESTED** [review]

File: src/librustdoc/clean/types.rs (hunk 21 above)
Comment (camelid) [reply src/librustdoc/clean/types.rs:1399]: I actually tried removing it, but I ran into some weird rustdoc behavior that made it trickier than it seemed at first. I'll open an issue.

Comment (camelid) [reply src/librustdoc/clean/types.rs:1399]: Opened #89287.

File: src/librustdoc/clean/mod.rs [hunk 24]
```diff
@@ -903,6 +903,12 @@ impl Clean<Type> for hir::TraitRef<'_> {
     }
 }
 
+impl Clean<Path> for hir::TraitRef<'_> {
+    fn clean(&self, cx: &mut DocContext<'_>) -> Path {
+        self.path.clean(cx)
+    }
+}
```
[the remark this answers is not in the document]
Comment (camelid) [reply src/librustdoc/clean/mod.rs]: I tried removing it, but I got the errors below, so it seems like it is used. You made this comment on an outdated diff, so maybe that's why it looked unused?

```
error[E0599]: no method named `clean` found for struct `rustc_hir::TraitRef` in the current scope
   --> src/librustdoc/clean/mod.rs:896:36
    |
896 |             trait_: self.trait_ref.clean(cx),
    |                                    ^^^^^ method not found in `rustc_hir::TraitRef<'_>`
    |
    = help: items from traits can only be used if the trait is implemented and in scope
note: `clean::Clean` defines an item `clean`, perhaps you need to implement it
   --> src/librustdoc/clean/mod.rs:55:1
    |
55  | crate trait Clean<T> {
    | ^^^^^^^^^^^^^^^^^^^^

error[E0599]: the method `clean` exists for enum `std::option::Option<rustc_hir::TraitRef<'_>>`, but its trait bounds were not satisfied
    --> src/librustdoc/clean/mod.rs:1871:33
     |
1871 |       let trait_ = impl_.of_trait.clean(cx);
     |                                   ^^^^^ method cannot be called on `std::option::Option<rustc_hir::TraitRef<'_>>` due to unsatisfied trait bounds
     |
     = note: the following trait bounds were not satisfied:
             `rustc_hir::TraitRef<'_>: clean::Clean<_>`
             which is required by `std::option::Option<rustc_hir::TraitRef<'_>>: clean::Clean<std::option::Option<_>>`
     = help: items from traits can only be used if the trait is implemented and in scope
note: `clean::Clean` defines an item `clean`, perhaps you need to implement it
    --> src/librustdoc/clean/mod.rs:55:1
     |
55   | crate trait Clean<T> {
     | ^^^^^^^^^^^^^^^^^^^^

For more information about this error, try `rustc --explain E0599`.
error: could not compile `rustdoc` due to 2 previous errors
```

File: src/librustdoc/clean/inline.rs (hunk 10 above)
Comment (camelid) [reply src/librustdoc/clean/inline.rs]: Because the `trait_` is no longer a `Type`; it's a `Path`.

File: src/librustdoc/clean/utils.rs (hunk 14 above)
Comment (camelid) [reply src/librustdoc/clean/utils.rs:162]: I guess not :)

It was actually only used in a single place before.

File: src/librustdoc/clean/utils.rs (hunk 15 above)
Comment (camelid) [reply src/librustdoc/clean/utils.rs:159]: Yeah, we should probably rename this function to something clearer. Maybe `strip_path_generics`?

File: src/librustdoc/html/render/mod.rs (hunk 16 above)
Comment (camelid) [reply src/librustdoc/html/render/mod.rs:2374]: Because it's now used in two places: here, and for `Type::QPath` below. `Type::QPath` can't just use `work` to do it because it's now a `Path` not a `Type`.

File: src/librustdoc/clean/inline.rs (hunk 18 above)
Comment (camelid) [reply src/librustdoc/clean/inline.rs:450]: `GetDefId::def_id` returns an `Option`, though, but we can always get the `DefId` of a `Path`. It adds unnecessary confusion about what information is available. I also find it very confusing that `GetDefId` is implement for `Option<T: GetDefId>`. This code is much clearer IMO.

File: src/librustdoc/clean/types.rs (hunk 22 above)
Comment (camelid) [reply src/librustdoc/clean/types.rs:1417]: > When can this happen? Shouldn't all types be inferred by the time rustdoc gets to them?

I don't know. Probably yes, since we don't even look at item bodies.

> so maybe this should say "a type error" instead and be removed?

What do you mean by "this should \[...\] be removed"? Either way, let's make this change in a separate PR so this one doesn't grow :)

File: src/librustdoc/clean/mod.rs (hunk 11 above)
Comment (camelid) [reply src/librustdoc/clean/mod.rs]: It uses `resolve_type`, which calls `register_res`. Also, it hasn't been changed in this PR.

File: src/librustdoc/clean/simplify.rs (hunk 12 above)
Comment (jyn514) [reply src/librustdoc/clean/simplify.rs]: :+1: yup that seems fine

File: src/librustdoc/html/render/mod.rs (hunk 17 above)
Comment (camelid) [reply src/librustdoc/html/render/mod.rs]: It looks like it will changes the order of the `Vec<String>` returned by the function this code is in (`collect_paths_for_type`). The order seems to be arbitrary though.

**Comment by camelid** [comment]
Unresolved review threads:

- [ ] https://github.com/rust-lang/rust/pull/88379#discussion_r697455410
- [ ] https://github.com/rust-lang/rust/pull/88379#discussion_r697463052
- [ ] https://github.com/rust-lang/rust/pull/88379#discussion_r697465028
- [ ] https://github.com/rust-lang/rust/pull/88379#discussion_r716253925
- [ ] https://github.com/rust-lang/rust/pull/88379#discussion_r716254124
- [ ] https://github.com/rust-lang/rust/pull/88379#discussion_r716254537
- [ ] https://github.com/rust-lang/rust/pull/88379#discussion_r716254748
- [ ] https://github.com/rust-lang/rust/pull/88379#discussion_r716255029
- [ ] https://github.com/rust-lang/rust/pull/88379#discussion_r716255173
- [ ] https://github.com/rust-lang/rust/pull/88379#discussion_r716255936
- [ ] https://github.com/rust-lang/rust/pull/88379#discussion_r716256743

**Comment by camelid** [comment]
Let's see what effect this has on perf.

@bors try @rust-timer queue

**Comment by camelid** [comment]
max-rss is significantly improved on several benchmarks too, with no regressions! 🎉

File: src/librustdoc/clean/types.rs (hunk 22 above)
Comment (jyn514) [reply src/librustdoc/clean/types.rs:1414]: I mean remove the `InferTy` variant of GenericArg altogether.

Yeah, another PR sounds fine.

File: src/librustdoc/clean/utils.rs (hunk 15 above)
Comment (jyn514) [reply src/librustdoc/clean/utils.rs:159]: :+1: seems useful

File: src/librustdoc/html/render/mod.rs (hunk 17 above)
Comment (jyn514) [reply src/librustdoc/html/render/mod.rs]: :+1: looks like this is only used in https://github.com/rust-lang/rust/blob/5f744f33f60f1961e756adf30606a08076350182/src/librustdoc/html/render/mod.rs#L1632, which is order independent.

**Comment by jyn514** [comment]
r=me with https://github.com/rust-lang/rust/pull/88379#discussion_r717825432 done :)

File: src/librustdoc/html/render/mod.rs (hunk 17 above)
Comment (camelid) [reply src/librustdoc/html/render/mod.rs]: I think it's also used when generating `implementors.js` files, but I think the order shouldn't matter there either.

**Comment by camelid** [comment]
By the way, I ran `x doc --stage 2 library/test` on the HEAD and the merge-base of this PR to ensure there weren't any changes. The only changes I noticed when scanning through `git diff --no-index -- old-doc new-doc` were re-ordering of some paths in the search index.

**Comment by camelid** [comment]
@bors rollup=never

This has a large perf impact.

**Comment by camelid** [comment]
@bors retry

All builders except two have finished, and they are stuck on "install awscli" and "checkout submodules".

**Comment by camelid** [comment]
`@hkratz` suggests that this timeout may be due to #89609. Hopefully that'll be fixed soon, and this can get merged!
